src/reactive/cinder_backup_swift_handlers.py

```diff
@@ -1,31 +1,29 @@
+from __future__ import absolute_import
 import charms.reactive as reactive
-
-import charms_openstack.charm
+import charms_openstack
 import charm.openstack.cinder_backup_swift
 import charms.reactive.flags as flags
-
 from charms.reactive.relations import (
     endpoint_from_flag,
 )
 
 charms_openstack.charm.defaults.use_defaults(
     'charm.installed',
-    'update-status')
-# if config has been changed we need to re-evaluate flags
-# config.changed is set and cleared (atexit) in layer-basic
+    'upgrade-charm',
+)
 
 flags.register_trigger(when='config.changed', clear_flag='config.complete')
 flags.register_trigger(when='upgraded', clear_flag='config.complete')
 flags.register_trigger(when='endpoint.backup-backend.changed',
                        clear_flag='config.complete')
 
 
-@reactive.when_any('endpoint.backup-backend.available')
+@reactive.when_any('endpoint.backup-backend.joined', 'endpoint.backup-backend.changed')
 @reactive.when_not('config.complete')
```

Comment:
This is a bit concerning. The ``cinder-backup`` interface code should be handling the ``joined`` and ``changed`` flags, introspect the relation data and set the ``available`` flag when all the required data is present.

Reply:
A replacement for the OSContextGenerator approach might be something like:

@when_not_all('config.default.swift-user', 'config.default.swift-key', ...)

This will make sure the handler does not run until all of those configs has been set.

The mandatory config approach mentioned above will take care of setting the workload status so that the end user is made aware of missing configuration.

Reply:
https://github.com/openstack-charmers/charm-interface-cinder-backup/pull/3

Reply:
switched to using 
@reactive.when_any('backup-backend.available')